### PR TITLE
feat(providers): re-introduce Atlassian (with OAuth 2)

### LINKED ISSUE #11613: Atlassian Provider not working but its in the docs for v5
### What is the improvement or update you wish to see?

I would like to see Atlassian implemented in v5. I see it is in v4, maybe you guys just haven't gotten around to it. All good, but remove it as an option from your v5 docs when you search for OAuth providers.

### Is there any context that might help us understand?

These are the docs for v5.
https://authjs.dev/getting-started/authentication/oauth

You can search for Atlassian but it is not in the v5 packages.
https://github.com/nextauthjs/next-auth/tree/main/packages/core/src/providers

### Does the docs page already exist? Please link to it.

https://authjs.dev/getting-started/authentication/oauth

### PR DESCRIPTION
**Pull request by DevRish**
## ☕️ Reasoning

The Atlassian provider which was removed in [cb1a260](https://github.com/nextauthjs/next-auth/commit/cb1a260523d7b5c6a5af59b63f1e911a60a9edb1) because then it only supported OAuth1.0. But at present it also supports OAuth2.0. This PR brings back the Atlassian provider with required configurations for OAuth2.0.

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Fixes #11613 

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

File: apps/examples/nextjs/auth.ts [hunk 1]
```diff
@@ -53,6 +54,11 @@ const config = {
   adapter: UnstorageAdapter(storage),
   providers: [
     Apple,
+    Atlassian({
+      clientId: process.env.AUTH_ATLASSIAN_ID,
+      clientSecret: process.env.AUTH_ATLASSIAN_SECRET,
+      redirectUri: process.env.AUTH_ATLASSIAN_REDIRECT_URI,
+    }),
```
Comment (balazsorban44) [inline apps/examples/nextjs/auth.ts:61]:
```suggestion
    Atlassian,
```

We detect these automatically. Furthermore, redirectUri is not a valid value for a provider config.

File: packages/core/src/providers/atlassian.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      url: "https://auth.atlassian.com/authorize",
+      params: {
+        audience: "api.atlassian.com",
+        prompt: "consent",
+        client_id: options.clientId,
+        scope: "read:me",
+        redirect_uri: options.redirectUri,
+        state: "sample_state", // required field, so must have
+        response_type: "code",
+      },
```
Comment (balazsorban44) [inline packages/core/src/providers/atlassian.ts:122]:
```suggestion
      params: {
        audience: "api.atlassian.com",
        scope: "read:me",
        state: "sample_state", // required field, so must have
      },
```

The rest of the parameters are already supposed to be sent correctly.

File: packages/core/src/providers/atlassian.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+ *
+ * If you think you found a bug in the default configuration, you can [open an issue](https://authjs.dev/new/provider-issue).
+ *
+ * Auth.js strictly adheres to the specification and it cannot take responsibility for any deviation from
+ * the spec by the provider. You can open an issue, but if the problem is non-compliance with the spec,
+ * we might not pursue a resolution. You can ask for more help in [Discussions](https://authjs.dev/new/github-discussions).
+ */
+export default function Atlassian<P extends AtlassianProfile>(
+  options: OAuthUserConfig<P> & AdditionalConfig
+): OAuthConfig<P> {
```
Comment (balazsorban44) [inline packages/core/src/providers/atlassian.ts:107]:
```suggestion
export default function Atlassian(
  options: OAuthUserConfig<AtlassianProfile>
): OAuthConfig<AtlassianProfile> {
```

File: packages/core/src/providers/atlassian.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+   */
+  picture: string
+}
+
+// Any config required that isn't part of the `OAuthUserConfig` spec should belong here
+// For example, we must pass a `redirectUri` to the Atlassian API when requesting tokens, therefore we add it here
+interface AdditionalConfig {
+  redirectUri: string | undefined
+}
+
```
Comment (balazsorban44) [inline packages/core/src/providers/atlassian.ts:40]:
```suggestion
```

This should not be needed

File: packages/core/src/providers/atlassian.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+ * export default NextAuth({
+ *   providers: [
+ *     AtlassianProvider({
+ *       clientId: process.env.ATLASSIAN_ID,
+ *       clientSecret: process.env.ATLASSIAN_SECRET,
+ *       redirectUri: process.env.ATLASSIAN_REDIRECT_URI,
+ *     }),
+ *   ],
+ * })
+ * ```
```
Comment (balazsorban44) [inline packages/core/src/providers/atlassian.ts:66]:
```suggestion
 * ```ts
 * import Atlassian from "@auth/core/providers/atlassian"
 * ...
 * providers: [
 *  Atlassian({
 *    clientId: env.AUTH_ATLASSIAN_ID,
 *    clientSecret: env.AUTH_ATLASSIAN_SECRET,
 *  }),
 * ]
 * ...
 * ```
```

Let's use a framework-agnostic example here.

File: packages/core/src/providers/atlassian.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ * 6. Click Save changes
+ * 7. Select **Settings** in the left menu
+ * 8. Access and copy your app's **Client ID** and **Secret**
+ *
+ * Then, create a `.env` file in the project root add the following entries:
+ *
+ * ```
+ * ATLASSIAN_ID=<Client ID copied in step 8>
+ * ATLASSIAN_SECRET=<Secret copied in step 8>
+ * ATLASSIAN_REDIRECT_URI=<Callback URL of your application>
```
Comment (balazsorban44) [inline packages/core/src/providers/atlassian.ts:86]:
```suggestion
```

File: packages/core/src/providers/atlassian.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+ * 5. Enter the **Callback URL**: `https://{YOUR_DOMAIN}/api/auth/callback/atlassian`
+ * 6. Click Save changes
+ * 7. Select **Settings** in the left menu
+ * 8. Access and copy your app's **Client ID** and **Secret**
+ *
+ * Then, create a `.env` file in the project root add the following entries:
+ *
+ * ```
+ * ATLASSIAN_ID=<Client ID copied in step 8>
+ * ATLASSIAN_SECRET=<Secret copied in step 8>
```
Comment (balazsorban44) [inline packages/core/src/providers/atlassian.ts:85]:
```suggestion
 * ATLASSIAN_AUTH_ID=<Client ID copied in step 8>
 * ATLASSIAN_AUTH_SECRET=<Secret copied in step 8>
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thanks! I added some suggestions to simplify the code.

Let me know if that still works!

File: packages/core/src/providers/atlassian.ts (hunk 7 above)
Comment (DevRish) [reply packages/core/src/providers/atlassian.ts]: Did you mean `AUTH_ATLASSIAN_ID` and `AUTH_ATLASSIAN_SECRET` (as those are what is mentioned for the env file, and also that convention is followed for other providers as well)

I changed them to `AUTH_ATLASSIAN_ID` and `AUTH_ATLASSIAN_SECRET`

**Comment by DevRish** [comment]
@balazsorban44  Thanks for your review! It works fine even after removing the params and simplification that you recommended. I have made all the changes that you requested.

**Review by balazsorban44 — APPROVED** [review]

File: packages/core/src/providers/atlassian.ts (hunk 7 above)
Comment (balazsorban44) [reply packages/core/src/providers/atlassian.ts]: Yes, sorry

**Pull request merged**